rust/gitxetcore/src/git_integration/git_repo_plumbing.rs

```diff
@@ -295,62 +366,122 @@ pub fn list_files_from_repo(
+
+                    return Ok(vec![entry]);
+                }
+                Some(ObjectType::Tree) => (repo.find_tree(entry.id())?, &prefix), // continue to list the subtree
+                _ => return Ok(vec![]), // unrecognized kind, return empty list
             }
+        } else {
+            return Ok(vec![]);
         }
-        Some(ObjectType::Blob) => return Ok(vec![Path::new(root_path).to_owned()]), // this is a file, directly return it
-        Some(ObjectType::Tree) => (), // continue to list the subtree
-        _ => return Ok(vec![]),       // unrecognized kind, return empty list
-    }
-
-    // now root_path is a directory
-    // special case, root_path is "." and will not be found by get_path
-    let (subtree, ancestor) = if root_path == REPO_ROOT_PATH {
-        (tree, "")
-    } else {
-        (repo.find_tree(root_path_entry_id)?, root_path)
     };
 
     let mut list = vec![];
     if !recursive {
         // only the direct children of this root_path
         for entry in subtree.iter() {
-            if let Some(git2::ObjectType::Blob) = entry.kind() {
-                let file_name = entry.name().unwrap().to_owned();
-                let file_path = Path::new(ancestor).join(file_name);
-                list.push(file_path);
+            if let Some(loe) = ListObjectEntry::from_tree_entry(repo, prefix, &entry) {
+                list.push(loe);
             }
         }
     } else {
         // recursively list all children under this root_path
         subtree.walk(TreeWalkMode::PreOrder, |parent, entry| {
             if let Some(git2::ObjectType::Blob) = entry.kind() {
-                let file_name = entry.name().unwrap().to_owned();
-                let file_path = Path::new(ancestor).join(parent).join(file_name);
-                list.push(file_path);
+                let dir_prefix = PathBuf::from(prefix).join(parent);
+                if let Some(loe) =
+                    ListObjectEntry::from_tree_entry(repo, dir_prefix.to_str().unwrap(), &entry)
+                {
+                    list.push(loe);
+                }
             }
-
             TreeWalkResult::Ok
         })?;
     }
 
     Ok(list)
 }
 
+/// Return a list of objects at a specific tree object
+pub fn list_objects_at_tree_oid(
+    repo: &Arc<Repository>,
+    tree_oid: git2::Oid, // The oid of the tree representing that directory.
+) -> Result<Vec<ListObjectEntry>> {
+    // Returns a sorted list of dir entries, either blobs or directories, given by the tree oid
+    let tree = repo.find_tree(tree_oid)?;
+    let mut entries = Vec::new();
+
+    for entry in tree.iter() {
+        let object = entry.to_object(repo)?;
+
+        let (size, is_tree) = match object.kind() {
+            Some(ObjectType::Blob) => (object.as_blob().unwrap().size(), false),
+            Some(ObjectType::Tree) => (0, true),
+            _ => {
+                warn!("Git object entry {object:?} in tree {tree:?} has type other than blob or tree.");
```

Comment:
maybe lower to debug!

Reply:
In this case it should never happen...   So something weird is going on if this happens.

Reply:
Ah yeah warn! is fine